### PR TITLE
change default behavior in appsi to not care where variables live

### LINKED ISSUE #2888: APPSI Solver Error: cannot remove variable q2 - it is still being used by constraints or the objective
## Summary
APPSI solver will report the `cannot remove variable q2 - it is still being used by constraints or the objective` error if the variable is deleted and defined again.

### Steps to reproduce the issue

```
# example.py
from pyomo.environ import *
from pyomo.solvers.tests.models.QCP_simple import QCP_simple

QCP_model = QCP_simple()
QCP_model._generate_model()
model = QCP_model.model
opt = SolverFactory('appsi_gurobi') # or 'appsi_cplex'
result = opt.solve(QCP_model.model,tee=True)
print(result)

QCP_model.model.del_component('q2')
model.q2 = Var(bounds=(-2, None))
result = opt.solve(QCP_model.model,tee=True)
print(result)
```

### Error Message

```console
$ # Output message here, including entire stack trace, if available
Model has 4 quadratic constraints
Coefficient statistics:
  Matrix range     [1e+00, 1e+00]
  QMatrix range    [1e+00, 1e+00]
  Objective range  [1e+00, 1e+00]
  Bounds range     [2e-01, 2e+00]
  RHS range        [1e+00, 1e+00]
  QRHS range       [2e-01, 5e+00]
Presolve removed 0 rows and 3 columns
Presolve time: 0.00s
Presolved: 7 rows, 7 columns, 14 nonzeros
Presolved model has 2 second-order cone constraints
Ordering time: 0.00s

Barrier statistics:
 AA' NZ     : 2.100e+01
 Factor NZ  : 2.800e+01
 Factor Ops : 1.400e+02 (less than 1 second per iteration)
 Threads    : 1

                  Objective                Residual
Iter       Primal          Dual         Primal    Dual     Compl     Time
   0   2.43809524e+00  2.43809524e+00  5.43e-01 4.40e-01  2.15e-01     0s
   1   2.49156971e+00  2.64695156e+00  3.89e-02 4.84e-07  2.42e-02     0s
   2   2.48970091e+00  2.50659507e+00  4.28e-08 5.31e-13  1.88e-03     0s
   3   2.49131949e+00  2.49163010e+00  9.30e-10 6.92e-14  3.45e-05     0s
   4   2.49156171e+00  2.49156262e+00  1.56e-12 4.00e-12  1.01e-07     0s

Barrier solved model in 4 iterations and 0.00 seconds (0.00 work units)
Optimal objective 2.49156171e+00


Problem: 
- Lower bound: 2.4915617113321344
  Upper bound: 2.4915617113321344
  Number of objectives: 1
  Number of constraints: 0
  Number of variables: 0
  Sense: -1
Solver: 
- Status: ok
  Termination condition: optimal
  Termination message: TerminationCondition.optimal
Solution: 
- number of solutions: 0
  number of solutions displayed: 0

Traceback (most recent call last):
  File "/Users/pengzedong1/Code/test.py", line 37, in <module>
    result = opt.solve(QCP_model.model,tee=True)
  File "/Users/pengzedong1/Github/pyomo/pyomo/contrib/appsi/base.py", line 1554, in solve
    results: Results = super(LegacySolverInterface, self).solve(model)
  File "/Users/pengzedong1/Github/pyomo/pyomo/contrib/appsi/solvers/gurobi.py", line 397, in solve
    self.update(timer=timer)
  File "/Users/pengzedong1/Github/pyomo/pyomo/contrib/appsi/solvers/gurobi.py", line 1146, in update
    super(Gurobi, self).update(timer=timer)
  File "/Users/pengzedong1/Github/pyomo/pyomo/contrib/appsi/base.py", line 1338, in update
    self.remove_variables(old_vars)
  File "/Users/pengzedong1/Github/pyomo/pyomo/contrib/appsi/base.py", line 1181, in remove_variables
    raise ValueError(
ValueError: cannot remove variable q2 - it is still being used by constraints or the objective
```

### Information on your system

Pyomo version: 6.6.1.dev0
Python version: 3.9.10
Operating system: macOS Monterey Version 12.1
How Pyomo was installed (PyPI, conda, source): source
Solver (if applicable):appsi_cplex, appsi_gurobi

### PR DESCRIPTION
**Pull request by michaelbynum**
## Summary/Motivation:
This PR 
- changes the default behavior of appsi solver interfaces to match that of the new NL writer so that models can be solved even if using variables that live outside the block being solved.
- fixes a bug in `update` where an error was raised when removing variables that were used in the old objective but not the new objective.
- improves the tests in appsi to test both `only_child_vars=True` and `only_child_vars=False`. This is the flag that controls how variables are gathered (by traversing the model or from the active constraints/objective).

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/contrib/appsi/base.py [hunk 1]
```diff
@@ -1463,6 +1460,12 @@ def update(self, timer: HierarchicalTimer = None):
             self.set_objective(pyomo_obj)
         timer.stop('objective')
 
+        # this has to be done after the objective in case the
+        # old objective uses old variables
+        timer.start('vars')
+        self.remove_variables(old_vars)
+        timer.stop('vars')
```
Comment (emma58) [inline pyomo/contrib/appsi/base.py:1467]: Just to make sure I'm understanding, it has to be after constraints for the same reason, right? (It just already was.)

**Review by emma58 — APPROVED** [review]
This is very happy! :)

**Comment by michaelbynum** [comment]
> This is very happy! :)

Haha, good!

File: pyomo/contrib/appsi/base.py (hunk 1 above)
Comment (michaelbynum) [reply pyomo/contrib/appsi/base.py:1467]: I'll update the comment

**Pull request merged**